File: backend/app/schemas/ticket.py



# datetime is used in response models (created_at/updated_at)
from datetime import datetime

# Enum ensures only allowed status values are accepted by the API
from enum import Enum

# Optional means this value can be None (nullable)
from typing import Optional

# Pydantic validates request/response data
from pydantic import BaseModel


class TicketStatus(str, Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"


class TicketCreate(BaseModel):
    """
    This schema is for: "Create Ticket" request body.

    Client sends ONLY fields they can control.
    They do NOT send id, created_at, updated_at because DB generates those.
    """

     # Required fields
    title: str
    message: str

    # Default status if not provided
    status: TicketStatus = TicketStatus.OPEN

    # Optional fields (can be missing or None)
    category: Optional[str] = None
    urgency: Optional[str] = None
    assigned_to: Optional[str] = None

class TicketRead(BaseModel):
        """
        This schema is for: "Ticket response" (what API returns).

        Includes DB-generated fields like id and timestamps.
        """
        id: int
        title: str
        message: str
        status: TicketStatus
        category: Optional[str] = None
        urgency: Optional[str] = None
        assigned_to: Optional[str] = None
        created_at: datetime
        updated_at: datetime



        # This tells Pydantic: "You may read data from ORM objects (SQLAlchemy models)"
        # So you can return a Ticket model directly and it will convert to this schema.

        class Config:
            orm_mode = True


class TicketUpdate(BaseModel):
    # for now, only allow status change (simple)
    status: TicketStatus
        

File: frontend/src/types/ticket.ts
// What the backend expects (API contract)
export type TicketStatus = "open" | "in_progress" | "resolved";

// What you show to users (UI labels)
export type TicketStatusLabel = "Open" | "In Progress" | "Resolved";


// This describes what a full Ticket looks like when backend sends it back.
// Think: "this is the ticket object shape we will display in the UI".
export type Ticket = {
  id: number;              // ticket unique number created by the database
  title: string;           // ticket title
  message: string;         // ticket message/details
  status: TicketStatus;    // must be Open/In Progress/Resolved
  category?: string | null;    // optional extra info (can be empty)
  urgency?: string | null;     // optional extra info (can be empty)
  assigned_to?: string | null; // optional extra info (can be empty)
  created_at: string;      // when it was created (from backend)
  updated_at: string;      // when it was last updated (from backend)
};

// This describes what we send when we create a new ticket.
// Notice: no id, created_at, updated_at because backend will generate them.
export type TicketCreate = {
  title: string;               // required
  message: string;             // required
  status?: TicketStatus;       // optional (if not sent, backend can default)
  category?: string | null;    // optional
  urgency?: string | null;     // optional
  assigned_to?: string | null; // optional
};